### ISSUE TITLE
[Post-v0.1][Taiwan][Data] NDC business-cycle indicators and CIER manufacturing PMI ingestion

### ISSUE THREAD

**Issue by world4jason** [world4jason/Market-Risk-Monitor#45]
> **Not a v0.1 blocker.** The first data-backed release in #46 can ship with TAIEX/CBC and the core U.S. history while NDC/CIER ingestion is completed separately.

## Objective

Decide how NDC business-cycle indicators and CIER manufacturing PMI get into the pipeline. This is the last unexecuted path in #42.

`docs/taiwan-sources.md` says these sources have "no guaranteed stable machine API" and accepts a manual normalized CSV via `--taiwan-macro-file`. I probed both to find out how true that still is. The answer differs per source, so they should be handled differently.

### CIER manufacturing PMI — automatable, but only a 12-month window

`https://www.cier.edu.tw/pmi-trend/` returns **200** and the PMI table is rendered server-side into static HTML. It parses cleanly:

```
月份     臺灣製造業PMI  化學暨生技醫療  電子暨光學  食品暨紡織  基礎原物料  交通工具  電力暨機械設備
2026/08  62.5％         54.5％          65.2％      52.4％      53.0％      50.6％    61.1％
2026/07  61.5％         51.0％          65.5％      58.8％      50.5％      59.5％    59.2％
...
2025/09  48.3％         48.3％          48.8％      41.3％      43.5％      41.1％    50.8％
```

Headline PMI plus all six sector series, exactly the `tw_manufacturing_pmi` canonical id the docs specify.

**The catch: the table holds 13 rows — a header and 12 months.** It is a rolling window, not a history. It is a JetEngine dynamic-table widget; deeper history would mean calling an undocumented Elementor `admin-ajax.php` endpoint, which is precisely the kind of unstable surface the docs warned about.

So a parser gets you a *fresh current value and a forward-accumulating history*, not a backfill.

### NDC business-cycle indicators — blocked from automation

Every NDC endpoint returns **403** from this network, with and without a browser User-Agent:

```
https://index.ndc.gov.tw/n/json/data/eco                        → 403
https://index.ndc.gov.tw/                                       → 403
https://www.ndc.gov.tw/Content_List.aspx?n=4CE0A9EFC9985010     → 403
```

The 403 bodies are edge/CDN protection pages, not "not found". `data.gov.tw` API forms I tried returned 405/404.

I did not attempt to work around the protection. A bot check is a statement about automated access, and routing around it to scrape a statistics page is not something I will do unprompted — say so explicitly if you want it attempted a different way.

## Proposal

Split the two, because they are not the same problem.

**CIER → build a parser.** It follows a pattern this repo already has: `scripts/bootstrap_cbc_rates.py` parses CBC's official HTML table into the normalized CSV contract, and that works well. A `bootstrap_cier_pmi.py` doing the same is straightforward, with two constraints stated up front:
- coverage starts at whatever the rolling window holds on first run and grows forward;
- `tw_manufacturing_pmi` gets an expected-history-start entry only once enough sessions accumulate, so the truncation guard does not fire spuriously on day one.

**NDC → manual normalized CSV, as the docs already specify.** Someone downloads the monitoring-indicator and index series and normalizes to:

```csv
date,provider,series_id,value,unit,release_date,source_url
2026-07-01,NDC,tw_ndc_monitoring_score,41,score,2026-08-27,https://www.ndc.gov.tw/en/
```

That file then feeds `--taiwan-macro-file`. Commit the normalized CSV so the ingest is reproducible even though the fetch is not.

## Acceptance Criteria

- [ ] `scripts/bootstrap_cier_pmi.py` fetches and normalizes the official CIER table into the documented CSV contract.
- [ ] Parser tests use a committed fixture of the real table shape and need no network.
- [ ] Percent signs and the `％` full-width character are handled; a malformed or empty table is rejected rather than parsed as zero.
- [ ] The 12-month rolling-window limitation is recorded in `docs/taiwan-sources.md`.
- [ ] A manual NDC snapshot is normalized, committed and ingested via `--taiwan-macro-file`.
- [ ] `scripts/validate_data.py` passes on the resulting `tw_manufacturing_pmi` and `tw_ndc_*` artifacts.
- [ ] Taiwan macro regime is generated from real inputs, and stale/missing macro inputs still resolve to `unknown` rather than a benign default.

## Notes

I did not synthesise a macro CSV to make the pipeline run green. Fabricating observations for a market-risk dashboard would be worse than leaving the gap visible, and #11's whole premise is that a broken fetch must never become a false signal.

Everything else in #42 now runs against real sources — TWSE, CBC, TAIEX (6868 obs) and the 1088-stock trend panel. This is the remaining piece.

**Comment by world4jason** [comment]
CIER half implemented in #47 — `pipeline/cier_pmi.py` + `scripts/bootstrap_cier_pmi.py`, feeding the existing `--taiwan-macro-file` contract. Verified live: 12 months 2025-09 → 2026-08, latest 62.5, and `tw_manufacturing_pmi.json` / `taiwan-macro-regime.json` / `taiwan-macro-audit.json` all pass `validate_data.py`.

Worth noting for this ticket's last AC: with PMI as the only available component the regime **refuses to score** — `known_components: 1` against a minimum of 2, `score: null`, `current: null`, and each missing NDC/production component carries an explicit `"vote": null`. Unknown, not a benign default, on real data.

Remaining here is the NDC manual snapshot. The 403s are unchanged.

**Comment by whiletrue247** [comment]
One potentially useful official route for the remaining NDC snapshot: Taiwan's Government Open Data Platform currently exposes NDC dataset **「景氣指標及燈號」** at https://data.gov.tw/dataset/6099. I rechecked the original page on 2026-09-22: it lists leading/coincident/lagging indices (including detrended series), the business-cycle signal composite score and signal, says the dataset updates monthly, is free under the Government Data Open License v1, and provides a ZIP resource. The page also explicitly warns that the historical leading/coincident/lagging series are revised retrospectively on monthly publication.

This does **not** prove the ZIP is fetchable from your CI/runtime or that its file/schema matches the repo contract, so I would still verify the binary download, historical coverage, encoding/schema, and revision handling before replacing the manual-snapshot path. But it looks more specific than trying generic data.gov.tw API forms and may be worth testing against the remaining NDC AC.

**Comment by whiletrue247** [comment]
One more concrete check on the official NDC route: I followed the ZIP resource link from data.gov.tw dataset 6099 again on 2026-09-22. The resource endpoint currently responds as `application/x-zip-compressed` rather than the 403 seen on the generic routes, so the file endpoint itself appears reachable from at least this environment. I still could not inspect the archive contents here because this client does not render ZIP binaries, so CSV/XLS filename(s), encoding, history depth, schema, and revision-handling still need local/CI verification before using it as the snapshot source. Official dataset page: https://data.gov.tw/dataset/6099

**Comment by world4jason** [comment]
Thanks — dataset 6099 is a much better lead than the generic API forms I tried. I verified the dataset page independently. Two of your points check out and matter more than the fetchability question; the fetchability itself did **not** reproduce here.

## Confirmed from the dataset page

```
提供機關    國家發展委員會
更新頻率    每1月
授權方式    政府資料開放授權條款-第1版
計費方式    免費
上架日期    2013-04-03
```

**The licence is the most valuable part of this.** Government Data Open License v1, free — that makes NDC data redistributable, which directly unblocks part of #44. Every other Taiwan/US source in the coverage table is official or public; the only genuinely unresolved one is the TraderMonty/FMP breadth series. NDC landing in the "clearly redistributable" column is worth having settled.

The ZIP resource resolves to:

```
filename  景氣指標及燈號.zip
path      /001/administrator/10/relfile/5781/6392/ea235bd9-...-d5c785d3d0e2.zip
host      ws.ndc.gov.tw
```

Note the host: `ws.ndc.gov.tw`, not the `index.ndc.gov.tw` / `www.ndc.gov.tw` hosts that 403'd earlier. That distinction is the actual substance of your first comment.

## ⚠️ The retroactive-revision warning is the real problem

You flagged this and it deserves to be the headline, not a footnote. Verbatim from the page:

> 每月發布時回溯修正領先、同時及落後指標的歷史資料。

Every monthly publication retrospectively revises the historical leading, coincident and lagging series.

That is a correctness problem for this repo's core premise, not a data-plumbing detail. `docs/data-contract.md` requires observation date and release date to be distinct precisely so that no-look-ahead analysis uses the release date. A series whose *history* is rewritten monthly means:

- today's `tw_ndc_leading_index` value for 2024-03 is **not** the value that was visible in 2024-03;
- ingesting the latest ZIP and treating its history as point-in-time would silently backfill revised figures into historical evaluation — the exact class of error #29 and #26 exist to prevent on the breadth side;
- the only honest options are to snapshot each monthly vintage and keep them, or to mark the series non-point-in-time and bar it from historical/event views, the way the MA-breadth series is already gated.

The monitoring score / signal (`tw_ndc_monitoring_score`) may be better behaved than the detrended indices, but that needs checking rather than assuming.

I would rather this were designed before any ingestion code is written, so I have not started one.

## ❌ The ZIP does not fetch from here

Your second comment saw `application/x-zip-compressed`. I get a 403, and the response headers say why:

```
HTTP/2 403
server: cloudflare
set-cookie: __cf_bm=...; Domain=ndc.gov.tw
cf-ray: a3eebb205e3c0f9a-TPE
content-type: text/html; charset=UTF-8
```

`__cf_bm` is Cloudflare's bot-management cookie, so this is environment-dependent bot filtering rather than a dead endpoint — consistent with you reaching it and me not. I did not attempt to work around it. A bot check is a statement about automated access, and engineering past one to scrape a statistics page is not something I will do on my own initiative.

Practical consequence: even if the archive is fine, **this cannot be assumed fetchable from an arbitrary runner**, which is the same constraint that made `docs/taiwan-sources.md` specify a manual normalized CSV in the first place.

## Suggested split

1. **Licence finding → #44.** NDC is Government Data Open License v1. That is a concrete answer for one source.
2. **Revision handling → design decision here, before code.** Either snapshot monthly vintages, or mark `tw_ndc_*` non-point-in-time and gate it out of historical views. This is the blocking question.
3. **Fetch path → treat as opportunistic.** If someone can download the ZIP from an environment that reaches it, commit the normalized CSV so the *ingest* is reproducible even where the *fetch* is not. That is what the CIER adapter in #47 already does for a source that is reachable.

If you can pull the archive from your side, the useful next artifact is its file listing, encoding and the earliest date present — that tells us whether vintage-snapshotting is even practical.

Meanwhile the CIER half is implemented and verified live in #47.

**Comment by world4jason** [comment]
Review of #47 surfaced a correctness bug in the last AC here, now fixed in `416d1e0`.

`build_macro_regime()` set `current` by walking back to the last non-unknown row, so a period with missing components kept showing the previous period's regime — and `assets/app.js` reads that field directly. A fully known June followed by a PMI-only July rendered as June's `expansion`. `pipeline/validate.py` enforced the same wrong rule, and the UI printed `Number(null).toFixed(2)` as a fabricated `score 0.00`.

`current` is now the latest period including unknown, the previous reading is kept as `latest_known`, and the UI shows the input count instead of inventing a score. So this ticket's *"stale/missing macro inputs still resolve to unknown"* AC is only now genuinely satisfied — my earlier comment claiming it on the single-component case covered the weaker "never had a known regime" path only.

**Comment by world4jason** [comment]
Progress update after superseding stale PR #47:

**CIER portion is now implemented on current main via #71.**
- official CIER PMI parser/bootstrap
- rolling 12-month source window accumulates forward rather than truncating history
- release_date is the verified-public upper bound
- first-ingest 12-month batch remains one availability batch under the release-aware PIT contract
- PMI-only macro regime remains unknown with missing configured components explicit in provenance
- 20 targeted CIER tests + full 225-test suite + offline bootstrap→refresh E2E passed

Related Taiwan macro stale-current correctness was separately fixed in #70.

**#45 remains open.** The remaining AC is the NDC side: normalized real snapshot / ingestion plus the revision-vintage policy for retrospectively revised NDC history.

**Comment by world4jason** [comment]
Reopened after independent review: #45 was closed while its NDC acceptance criteria were still incomplete.

CIER ingestion is implemented and its immediate correctness follow-ups are in PR #72, but #45 still owns:
- a real normalized NDC snapshot / ingestion path;
- explicit revision/vintage handling for retrospectively revised NDC history;
- the combined macro assembly contract so CIER and NDC can be refreshed without one source replacing/pruning the other.

The last item matters because current `refresh_taiwan_macro_file()` treats one CSV as the full macro dataset. The NDC implementation should resolve that composition contract rather than merely add a second downloader.

**Comment by world4jason** [comment]
Next implementation is open as PR #73: https://github.com/world4jason/Market-Risk-Monitor/pull/73

This does not close #45. It addresses the multi-source composition trap first:
- repeatable `--taiwan-macro-file`
- deterministic CIER + NDC union
- one source owner per canonical series
- partial refresh cannot omit an already published series
- partial refresh cannot truncate previously published dates
- guard runs before writes, so a CIER-only follow-up cannot silently erase NDC outputs
- NDC remains current-vintage / retrospective / non-PIT

Still required for #45 completion: obtain a real official NDC snapshot by an allowed/manual path, normalize/commit it, then ingest it through this assembly contract. No NDC observations are synthesized in #73.
